### DISCUSSION TITLE
Duplicate and extend an extension

### DISCUSSION THREAD

**Discussion by Mushr0000m** [ueberdosis/tiptap#2031] [General Questions] [answered]
Hi,

I want to create a button that create a Paragraph Node with a specific class, let say `<p class="foo">...</p>`. So I read the doc https://tiptap.dev/guide/custom-extensions#attributes and created a CustomParagraph for my tests. Like this (that color part is not relevant is just for testing new extension) : 
```
import Paragraph from '@tiptap/extension-paragraph';
export const ReferenceParagraph = Paragraph.extend({
    addAttributes() {
        return {
            color: {
                default: 'pink',
            },
        };
    },
});
```
then :
```
editor = new Editor({
    // other code
    extensions: [
        StarterKit,
        ReferenceParagraph,
    ]
});
```

But my issue is doing this replace the default paragraph node and does not create a new type of node. Can someone explain how to do this, I don't understand the way to go ? Keeping the Paragraph node + adding and other one that does the same thing with custom class ?

Thanks

**Comment by cadars** [comment, marked as answer]
I think you need to name the custom extension: `name: 'ReferenceParagraph',` to differentiate it from the extended one (and perhaps rename/redefine the commands and keyboard shortcuts as well)

**Reply by Mushr0000m** [reply]
Ok I'll have a look a this thnks ! And do you know how to add the class ? I only see examples with attributes other that classes.

**Reply by Mushr0000m** [reply]
Ok so I made it work to create a custom paragraph with a custom class : 
```
export const customParagraph = Node.create({
    name: 'customParagraph',
    priority: 1001,
    group: 'block',
    content: 'inline*',
    parseHTML() {
        return [
            { tag: 'p.custom' },
        ];
    },
    renderHTML() {
        return ['p', {class: 'custom'}, 0];
    },
    addCommands() {
        return {
            toggleCustomParagraph: () => ({ commands }) => {
                return commands.toggleNode(this.name, 'paragraph');
            },
        };
    },
});
```

**Reply by Mushr0000m** [reply]
So.... it's working BUT I can't make a normal `p` now. Can it be from the `parseHTML` ? I had to put a higher priority than the Paragraph extension because if not, the class was remove when refreshing the page. I though it was because the Paragraph node would match before mine because it's simply a `p` tag. So now if at the end of my custom paragraph I hit enter key on the keyboard it creates a news custom paragraph and not a basic paragraph...

**Reply by cadars** [reply]
Have you tried to add the class via `HTMLAttributes` instead? Looks like that’s the recommended way to do it.

```js
defaultOptions: {
  ...Paragraph.options, // not sure about this line
  HTMLAttributes: {
    class: 'custom',
  },
},
```

**Reply by Mushr0000m** [reply]
I tried it but it's the same. The issue seems to be from the fact that I use a `p` tag which is the same tag as the default node tag, so if the priority is lower than the Paragraph extension, mine is not getting called when refreshing the page. The scenario is:

1. I type "something" in the editor : `<p>something</p>`
2. Click on my custom paragraph button
3. The sentence is wrap becoming `<p class="custom">something</p>`
4. Behind the scene it's saved in my DB (I checked and everything is fine in it)
5. I refresh the page and `<p class="custom">something</p>` is passed to the editor as content
6. The content revert to `<p>something</p>` so losing the class

I think the Paragraph extension is triggered first

**Reply by Mushr0000m** [reply]
I found it ! I read the ProseMirror doc and found there is an other `priority` attribute on the schema, so the key was to remove mine from the Node and add the one on the schema with higher value than the Paragraph extension, so my extension match before the Paragraph one but leave the Paragraph as the default `p` tag.

```
export const customParagraph = Node.create({
    name: 'customParagraph',
    group: 'block',
    content: 'inline*',
    parseHTML() {
        return [
            {
                tag: 'p.custom',
                priority: 200,
             },
        ];
    },
    renderHTML() {
        return ['p', {class: 'custom'}, 0];
    },
    addCommands() {
        return {
            toggleCustomParagraph: () => ({ commands }) => {
                return commands.toggleNode(this.name, 'paragraph');
            },
        };
    },
});
```

Hope it'll help others 😉
